src/frequenz/sdk/power/_distribution_algorithm.py

```diff
@@ -336,57 +346,98 @@ def _distribute_power(
+            components, available_soc, excl_bounds
+        )
 
         distribution: Dict[int, float] = {}
-
+        print(f"{power_w=}")
         # sum_ratio == 0 means that all batteries are fully charged / discharged
         if is_close_to_zero(sum_ratio):
             distribution = {inverter.component_id: 0 for _, inverter in components}
             return DistributionResult(distribution, power_w)
 
         distributed_power: float = 0.0
+        reserved_power: float = 0.0
         power_to_distribute: float = power_w
         used_ratio: float = 0.0
         ratio = sum_ratio
-        for pair, battery_ratio in battery_availability_ratio:
+        excess_reserved: dict[int, float] = {}
+        deficits: dict[int, float] = {}
+        for pair, excl_bound, battery_ratio in battery_availability_ratio:
             inverter = pair[1]
             # ratio = 0, means all remaining batteries reach max SoC lvl or have no
             # capacity
             if is_close_to_zero(ratio):
                 distribution[inverter.component_id] = 0.0
                 continue
 
-            distribution[inverter.component_id] = (
-                power_to_distribute * battery_ratio / ratio
-            )
-
+            power_to_distribute = power_w - reserved_power
+            calculated_power = power_to_distribute * battery_ratio / ratio
+            reserved_power += max(calculated_power, excl_bound)
             used_ratio += battery_ratio
-
+            ratio = sum_ratio - used_ratio
             # If the power allocated for that inverter is out of bound,
             # then we need to distribute more power over all remaining batteries.
-            upper_bound = upper_bounds[inverter.component_id]
-            if distribution[inverter.component_id] > upper_bound:
-                distribution[inverter.component_id] = upper_bound
-                distributed_power += upper_bound
-                # Distribute only the remaining power.
-                power_to_distribute = power_w - distributed_power
-                # Distribute between remaining batteries
-                ratio = sum_ratio - used_ratio
+            incl_bound = incl_bounds[inverter.component_id]
+            if calculated_power > incl_bound:
+                excess_reserved[inverter.component_id] = incl_bound - excl_bound
+                # # Distribute between remaining batteries
+            elif calculated_power < excl_bound:
+                deficits[inverter.component_id] = calculated_power - excl_bound
             else:
-                distributed_power += distribution[inverter.component_id]
+                excess_reserved[inverter.component_id] = calculated_power - excl_bound
+
+            distributed_power += excl_bound
+            distribution[inverter.component_id] = excl_bound
+
+        for inverter_id, deficit in deficits.items():
+            while not math.isclose(deficit, 0.0, abs_tol=1e-6) and deficit < 0.0:
```

Comment:
You could use `._internal._math.is_close_to_zero()` here.
```suggestion
            while not is_close_to_zero(deficit) and deficit < 0.0:
```